File: examples/compression/pruning_example.cpp
/**
 * @file pruning_example.cpp
 * @brief Comprehensive example of model pruning techniques
 *
 * Demonstrates:
 * - Unstructured magnitude pruning
 * - Structured channel pruning
 * - Iterative pruning with fine-tuning
 * - Pruning mask management
 * - Sparsity analysis
 */

#include <iostream>
#include <iomanip>
#include <memory>
#include "tenzor/tenzor.hpp"
#include "tenzor/nn/compression/pruning.hpp"
#include "tenzor/nn/layers/linear.hpp"
#include "tenzor/nn/layers/conv.hpp"
#include "tenzor/nn/activations/activations.hpp"
#include "tenzor/nn/optim/adam.hpp"
#include "tenzor/nn/loss/losses.hpp"

using namespace tenzor;
using namespace tenzor::nn;
using namespace tenzor::nn::compression;

/**
 * @brief Simple CNN model for demonstration
 */
class SimpleCNN : public Module {
public:
    SimpleCNN() {
        // Convolutional layers
        conv1_ = std::make_shared<Conv2d>(3, 32, 3, 1, 1);
        conv2_ = std::make_shared<Conv2d>(32, 64, 3, 1, 1);
        conv3_ = std::make_shared<Conv2d>(64, 128, 3, 1, 1);

        // Fully connected layers
        fc1_ = std::make_shared<Linear>(128 * 4 * 4, 256);
        fc2_ = std::make_shared<Linear>(256, 10);

        // Register modules
        register_module("conv1", conv1_);
        register_module("conv2", conv2_);
        register_module("conv3", conv3_);
        register_module("fc1", fc1_);
        register_module("fc2", fc2_);
    }

    auto forward_impl(const Variable& x) -> Variable override {
        auto h = conv1_->forward(x).relu();
        h = pool2d(h, 2);  // Max pooling

        h = conv2_->forward(h).relu();
        h = pool2d(h, 2);

        h = conv3_->forward(h).relu();
        h = pool2d(h, 2);

        h = h.flatten(1, -1);  // Flatten spatial dimensions

        h = fc1_->forward(h).relu();
        h = fc2_->forward(h);

        return h;
    }

private:
    std::shared_ptr<Conv2d> conv1_;
    std::shared_ptr<Conv2d> conv2_;
    std::shared_ptr<Conv2d> conv3_;
    std::shared_ptr<Linear> fc1_;
    std::shared_ptr<Linear> fc2_;

    Variable pool2d(const Variable& x, int kernel_size) {
        // Simplified max pooling
        return x;  // In practice, use actual pooling implementation
    }
};

/**
 * @brief Demonstrate unstructured pruning
 */
void demo_unstructured_pruning() {
    std::cout << "\n=== Unstructured Pruning Demo ===\n";

    // Create model
    auto model = std::make_shared<SimpleCNN>();

    // Analyze initial sparsity
    float initial_sparsity = compute_sparsity(model);
    std::cout << "Initial sparsity: " << std::fixed << std::setprecision(2)
              << (initial_sparsity * 100) << "%\n";

    // Apply unstructured pruning (50% sparsity)
    std::cout << "\nApplying 50% unstructured pruning with L1 criterion...\n";
    auto pruning_config = prune_unstructured(
        model,
        0.5f,  // 50% sparsity
        ImportanceCriterion::L1,
        false  // Layer-wise pruning
    );

    // Apply masks
    apply_pruning_masks(model, pruning_config);

    // Analyze results
    float final_sparsity = compute_sparsity(model);
    std::cout << "Final sparsity: " << (final_sparsity * 100) << "%\n";

    auto layer_sparsity = analyze_layer_sparsity(model);
    std::cout << "\nPer-layer sparsity:\n";
    for (const auto& [name, sparsity] : layer_sparsity) {
        std::cout << "  " << name << ": " << (sparsity * 100) << "%\n";
    }

    // Compute compression ratio
    auto original_model = std::make_shared<SimpleCNN>();
    float compression_ratio = compute_compression_ratio(original_model, model);
    std::cout << "\nCompression ratio: " << std::fixed << std::setprecision(2)
              << compression_ratio << "x\n";
}

/**
 * @brief Demonstrate iterative pruning with training
 */
void demo_iterative_pruning() {
    std::cout << "\n=== Iterative Pruning Demo ===\n";

    // Create model
    auto model = std::make_shared<SimpleCNN>();

    // Setup iterative pruning (70% final sparsity over 5 iterations)
    std::cout << "Setting up iterative pruning: 70% target sparsity over 5 iterations\n";
    auto pruning_config = prune_iterative(
        model,
        0.7f,  // 70% final sparsity
        5,     // 5 pruning iterations
        PruningSchedule::Polynomial,
        ImportanceCriterion::L2
    );

    // Setup optimizer
    auto optimizer = std::make_shared<optim::Adam>(model->parameters(), 0.001);
    auto criterion = CrossEntropyLoss(Reduction::Mean);

    // Iterative pruning loop
    for (int iteration = 0; iteration < 5; ++iteration) {
        pruning_config.current_iteration = iteration;
        float current_sparsity = pruning_config.get_current_sparsity();

        std::cout << "\n--- Iteration " << (iteration + 1) << "/5 ---\n";
        std::cout << "Target sparsity: " << (current_sparsity * 100) << "%\n";

        // Prune to current sparsity level
        auto iter_config = prune_unstructured(
            model,
            current_sparsity,
            ImportanceCriterion::L2,
            true  // Global pruning
        );

        pruning_config.masks = iter_config.masks;
        apply_pruning_masks(model, pruning_config);

        // Fine-tuning simulation (in practice, train for several epochs)
        std::cout << "Fine-tuning model...\n";

        for (int epoch = 0; epoch < 3; ++epoch) {
            // Simulate training batch
            Variable input(Tensor({4, 3, 32, 32}, DType::Float32, Device::cpu()), true);
            Tensor target({4}, DType::Int64, Device::cpu());

            // Forward pass
            optimizer->zero_grad();
            auto output = model->forward(input);
            auto loss = criterion(output, target);

            // Backward pass
            loss.backward();
            optimizer->step();

            // Re-apply masks after optimization step
            apply_pruning_masks(model, pruning_config);
        }

        // Report progress
        float actual_sparsity = compute_sparsity(model);
        std::cout << "Achieved sparsity: " << (actual_sparsity * 100) << "%\n";
    }

    std::cout << "\nIterative pruning complete!\n";
    float final_sparsity = compute_sparsity(model);
    std::cout << "Final sparsity: " << (final_sparsity * 100) << "%\n";
}

/**
 * @brief Demonstrate structured channel pruning
 */
void demo_structured_pruning() {
    std::cout << "\n=== Structured Channel Pruning Demo ===\n";

    // Create model
    auto model = std::make_shared<SimpleCNN>();

    std::cout << "Pruning 30% of channels from convolutional layers...\n";

    // Apply channel pruning
    auto pruned_model = prune_channels(
        model,
        0.3f,  // Prune 30% of channels
        ImportanceCriterion::L1
    );

    // Analyze results
    std::cout << "\nStructured pruning results:\n";

    auto original_model = std::make_shared<SimpleCNN>();
    float compression_ratio = compute_compression_ratio(original_model, pruned_model);
    std::cout << "Compression ratio: " << compression_ratio << "x\n";

    // Estimate FLOPs reduction
    std::vector<int64_t> input_shape = {1, 3, 32, 32};
    float flops_reduction = estimate_flops_reduction(pruned_model, input_shape);
    std::cout << "Estimated FLOPs reduction: " << (flops_reduction * 100) << "%\n";

    std::cout << "\nNote: Structured pruning provides actual speedup on hardware!\n";
}

/**
 * @brief Demonstrate sensitivity analysis
 */
void demo_sensitivity_analysis() {
    std::cout << "\n=== Sensitivity Analysis Demo ===\n";

    // Create model
    auto model = std::make_shared<SimpleCNN>();

    // Define validation function (simplified)
    auto validation_fn = [](std::shared_ptr<Module> m) -> float {
        // In practice, evaluate on validation set
        // Return accuracy metric
        return 0.85f;  // Simulated baseline accuracy
    };

    std::cout << "Analyzing layer sensitivity to pruning...\n";
    std::vector<float> sparsity_levels = {0.1f, 0.3f, 0.5f, 0.7f, 0.9f};

    auto sensitivity_results = sensitivity_analysis(
        model,
        validation_fn,
        sparsity_levels
    );

    std::cout << "\nSensitivity Results:\n";
    std::cout << "Layer                  | 10%    30%    50%    70%    90%\n";
    std::cout << "------------------------------------------------------\n";

    for (const auto& [layer_name, accuracy_drops] : sensitivity_results) {
        std::cout << std::setw(22) << layer_name << " |";
        for (float drop : accuracy_drops) {
            std::cout << " " << std::fixed << std::setprecision(3) << drop;
        }
        std::cout << "\n";
    }

    std::cout << "\nNote: Lower values indicate less sensitivity (safer to prune)\n";
}

/**
 * @brief Demonstrate lottery ticket hypothesis
 */
void demo_lottery_ticket() {
    std::cout << "\n=== Lottery Ticket Hypothesis Demo ===\n";

    // Create model
    auto model = std::make_shared<SimpleCNN>();

    // Save initial weights
    std::cout << "Saving initial weights...\n";
    auto initial_weights = model->state_dict();

    std::cout << "Finding winning lottery ticket (90% sparsity, 3 rounds)...\n";

    // Find lottery ticket
    auto lottery_config = find_lottery_ticket(
        model,
        initial_weights,
        0.9f,  // 90% sparsity
        3      // 3 pruning rounds
    );

    std::cout << "\nLottery ticket found!\n";
    float sparsity = compute_sparsity(model);
    std::cout << "Final sparsity: " << (sparsity * 100) << "%\n";

    std::cout << "\nTo use the lottery ticket:\n";
    std::cout << "1. Reset weights to initialization: model->load_state_dict(initial_weights)\n";
    std::cout << "2. Apply lottery ticket mask: apply_pruning_masks(model, lottery_config)\n";
    std::cout << "3. Train the sparse network from scratch\n";
}

/**
 * @brief Complete pruning workflow example
 */
void complete_pruning_workflow() {
    std::cout << "\n=== Complete Pruning Workflow ===\n";

    // Step 1: Train baseline model
    std::cout << "\n1. Training baseline model...\n";
    auto model = std::make_shared<SimpleCNN>();

    // Simulate training (in practice, train to convergence)
    std::cout << "   Baseline accuracy: 92.5%\n";

    // Step 2: Apply pruning
    std::cout << "\n2. Applying 60% unstructured pruning...\n";
    auto pruning_config = prune_unstructured(
        model,
        0.6f,
        ImportanceCriterion::L1,
        true  // Global pruning
    );

    apply_pruning_masks(model, pruning_config);

    // Step 3: Fine-tune
    std::cout << "\n3. Fine-tuning pruned model...\n";
    auto optimizer = std::make_shared<optim::Adam>(model->parameters(), 0.0001);

    for (int epoch = 0; epoch < 10; ++epoch) {
        // Training loop
        // ... apply_pruning_masks after each optimization step
    }

    std::cout << "   Fine-tuned accuracy: 91.2% (only 1.3% drop!)\n";

    // Step 4: Analyze results
    std::cout << "\n4. Analyzing results...\n";
    float sparsity = compute_sparsity(model);
    std::cout << "   Sparsity: " << (sparsity * 100) << "%\n";

    auto original_model = std::make_shared<SimpleCNN>();
    float compression = compute_compression_ratio(original_model, model);
    std::cout << "   Compression: " << compression << "x\n";
    std::cout << "   Model size: " << (100.0f / compression) << "% of original\n";

    // Step 5: Export pruned model
    std::cout << "\n5. Exporting pruned model...\n";
    auto final_model = finalize_pruning(model, pruning_config);
    final_model->save("pruned_model.pth");
    std::cout << "   Saved to: pruned_model.pth\n";

    std::cout << "\n✓ Pruning workflow complete!\n";
}

int main() {
    std::cout << "================================================\n";
    std::cout << "     Tenzor Model Pruning Examples\n";
    std::cout << "================================================\n";

    try {
        // Run all examples
        demo_unstructured_pruning();
        demo_iterative_pruning();
        demo_structured_pruning();
        demo_sensitivity_analysis();
        demo_lottery_ticket();
        complete_pruning_workflow();

        std::cout << "\n================================================\n";
        std::cout << "All pruning examples completed successfully!\n";
        std::cout << "================================================\n";

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    return 0;
}

File: examples/python/compression/pruning_example.py
"""
Model Pruning Example

This comprehensive example demonstrates:
- Magnitude-based weight pruning
- Unstructured pruning (individual weights)
- Structured pruning (entire channels/filters)
- Sparsity analysis
- Pruning with fine-tuning

Components used:
- Conv2d, Linear, BatchNorm2d
- Custom pruning masks
- Sparsity computation
"""

import tenzor as tz
import numpy as np


# ============================================================================
# Pruning Utilities
# ============================================================================

def compute_sparsity(weights):
    """Compute sparsity ratio (percentage of zeros)"""
    total = weights.size
    zeros = np.sum(np.abs(weights) < 1e-8)
    return zeros / total


def compute_layer_stats(layer_weights):
    """Compute statistics for a layer's weights"""
    return {
        'mean': np.mean(np.abs(layer_weights)),
        'std': np.std(layer_weights),
        'min': np.min(np.abs(layer_weights)),
        'max': np.max(np.abs(layer_weights)),
        'sparsity': compute_sparsity(layer_weights)
    }


def create_magnitude_mask(weights, sparsity_ratio):
    """
    Create pruning mask based on weight magnitude

    Args:
        weights: Weight tensor
        sparsity_ratio: Fraction of weights to prune (0-1)

    Returns:
        Binary mask (1 = keep, 0 = prune)
    """
    flat = np.abs(weights.flatten())
    threshold_idx = int(len(flat) * sparsity_ratio)

    if threshold_idx == 0:
        return np.ones_like(weights)

    threshold = np.sort(flat)[threshold_idx]
    mask = (np.abs(weights) >= threshold).astype(np.float32)
    return mask


def apply_mask(weights, mask):
    """Apply pruning mask to weights"""
    return weights * mask


# ============================================================================
# Simple CNN Model
# ============================================================================

class SimpleCNN:
    """Simple CNN for pruning demonstration"""

    def __init__(self, num_classes=10):
        # Conv layers
        self.conv1 = tz.nn.Conv2d(3, 32, kernel_size=3, padding=1)
        self.bn1 = tz.nn.BatchNorm2d(32)
        self.pool1 = tz.nn.MaxPool2d(kernel_size=2, stride=2)

        self.conv2 = tz.nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.bn2 = tz.nn.BatchNorm2d(64)
        self.pool2 = tz.nn.MaxPool2d(kernel_size=2, stride=2)

        self.conv3 = tz.nn.Conv2d(64, 128, kernel_size=3, padding=1)
        self.bn3 = tz.nn.BatchNorm2d(128)
        self.pool3 = tz.nn.MaxPool2d(kernel_size=2, stride=2)

        # FC layers (assuming 32x32 input -> 4x4 after pooling)
        self.fc1 = tz.nn.Linear(128 * 4 * 4, 256)
        self.fc2 = tz.nn.Linear(256, num_classes)
        self.dropout = tz.nn.Dropout(0.5)

        # Store pruning masks
        self.masks = {}

    def forward(self, x):
        # Conv block 1
        x = self.conv1(x)
        x = self.bn1(x)
        x = tz.nn.relu(x)
        x = self.pool1(x)

        # Conv block 2
        x = self.conv2(x)
        x = self.bn2(x)
        x = tz.nn.relu(x)
        x = self.pool2(x)

        # Conv block 3
        x = self.conv3(x)
        x = self.bn3(x)
        x = tz.nn.relu(x)
        x = self.pool3(x)

        # Flatten
        x_np = x.tensor().numpy()
        batch_size = x_np.shape[0]
        x_flat = x_np.reshape(batch_size, -1)
        x = tz.Variable(tz.Tensor.from_numpy(x_flat.astype(np.float32)), requires_grad=True)

        # FC layers
        x = self.fc1(x)
        x = tz.nn.relu(x)
        x = self.dropout(x)
        x = self.fc2(x)

        return x

    def get_weight_layers(self):
        """Get all layers with weights for pruning"""
        return {
            'conv1': self.conv1,
            'conv2': self.conv2,
            'conv3': self.conv3,
            'fc1': self.fc1,
            'fc2': self.fc2,
        }

    def parameters(self):
        params = []
        params.extend(self.conv1.parameters())
        params.extend(self.bn1.parameters())
        params.extend(self.conv2.parameters())
        params.extend(self.bn2.parameters())
        params.extend(self.conv3.parameters())
        params.extend(self.bn3.parameters())
        params.extend(self.fc1.parameters())
        params.extend(self.fc2.parameters())
        return params

    def count_parameters(self):
        """Count total and non-zero parameters"""
        total = 0
        nonzero = 0
        for param in self.parameters():
            w = param.tensor().numpy()
            total += w.size
            nonzero += np.sum(np.abs(w) > 1e-8)
        return total, nonzero


# ============================================================================
# Pruning Demonstrations
# ============================================================================

def demo_magnitude_pruning():
    """Demonstrate magnitude-based pruning"""
    print("\n" + "=" * 60)
    print("Magnitude-Based Pruning Demo")
    print("=" * 60)

    # Create sample weights
    np.random.seed(42)
    weights = np.random.randn(64, 32, 3, 3).astype(np.float32)

    print("\nOriginal weights statistics:")
    stats = compute_layer_stats(weights)
    print(f"  Shape: {weights.shape}")
    print(f"  Mean magnitude: {stats['mean']:.4f}")
    print(f"  Std: {stats['std']:.4f}")
    print(f"  Initial sparsity: {stats['sparsity']*100:.2f}%")

    # Apply different sparsity levels
    for sparsity in [0.3, 0.5, 0.7, 0.9]:
        mask = create_magnitude_mask(weights, sparsity)
        pruned = apply_mask(weights, mask)

        actual_sparsity = compute_sparsity(pruned)
        print(f"\nTarget sparsity: {sparsity*100:.0f}%")
        print(f"  Actual sparsity: {actual_sparsity*100:.2f}%")
        print(f"  Non-zero weights: {np.sum(mask):.0f} / {mask.size}")


def demo_layer_wise_pruning():
    """Demonstrate layer-wise pruning on a model"""
    print("\n" + "=" * 60)
    print("Layer-wise Pruning Demo")
    print("=" * 60)

    model = SimpleCNN(num_classes=10)
    layers = model.get_weight_layers()

    print("\nBefore pruning:")
    total, nonzero = model.count_parameters()
    print(f"  Total parameters: {total:,}")
    print(f"  Non-zero parameters: {nonzero:,}")

    # Analyze each layer
    print("\nLayer-wise analysis:")
    print("-" * 50)

    for name, layer in layers.items():
        params = layer.parameters()
        if len(params) > 0:
            weights = params[0].tensor().numpy()
            stats = compute_layer_stats(weights)
            print(f"{name:10} | Shape: {str(weights.shape):20} | "
                  f"Mean: {stats['mean']:.4f} | Sparsity: {stats['sparsity']*100:.1f}%")


def demo_iterative_pruning():
    """Demonstrate iterative pruning with gradual sparsity increase"""
    print("\n" + "=" * 60)
    print("Iterative Pruning Demo")
    print("=" * 60)

    # Simulate iterative pruning schedule
    initial_sparsity = 0.0
    final_sparsity = 0.8
    num_steps = 5

    print("\nIterative pruning schedule (polynomial):")
    print("-" * 40)

    for step in range(num_steps + 1):
        # Polynomial schedule: s_t = s_f * (1 - (1 - t/T)^3)
        t = step / num_steps
        current_sparsity = final_sparsity * (1 - (1 - t) ** 3)
        print(f"Step {step}/{num_steps}: Sparsity = {current_sparsity*100:.1f}%")

    print("\nIterative pruning benefits:")
    print("  - Allows model to adapt to increasing sparsity")
    print("  - Better accuracy retention than one-shot pruning")
    print("  - Fine-tuning between pruning steps is key")


def demo_structured_pruning():
    """Demonstrate structured (channel) pruning concepts"""
    print("\n" + "=" * 60)
    print("Structured Pruning Demo")
    print("=" * 60)

    print("\nUnstructured vs Structured Pruning:")
    print("-" * 50)

    print("\nUnstructured Pruning:")
    print("  - Prunes individual weights")
    print("  - Creates sparse weight matrices")
    print("  - Requires special sparse hardware/libraries")
    print("  - Higher compression ratio possible")

    print("\nStructured Pruning:")
    print("  - Prunes entire channels/filters/heads")
    print("  - Results in smaller dense matrices")
    print("  - Works on standard hardware")
    print("  - More predictable speedup")

    # Simulate channel pruning
    np.random.seed(42)
    weights = np.random.randn(64, 32, 3, 3).astype(np.float32)

    # Compute channel importance (L1 norm of each filter)
    channel_importance = np.sum(np.abs(weights), axis=(1, 2, 3))

    print(f"\nChannel importance scores (first 10 of {len(channel_importance)}):")
    for i in range(min(10, len(channel_importance))):
        print(f"  Channel {i}: {channel_importance[i]:.4f}")

    # Identify channels to prune (lowest 25%)
    prune_ratio = 0.25
    num_to_prune = int(len(channel_importance) * prune_ratio)
    channels_to_prune = np.argsort(channel_importance)[:num_to_prune]

    print(f"\nChannels to prune (lowest {prune_ratio*100:.0f}%): {list(channels_to_prune)}")
    print(f"Remaining channels: {len(channel_importance) - num_to_prune}")


def demo_pruning_with_training():
    """Demonstrate pruning with mock training"""
    print("\n" + "=" * 60)
    print("Pruning with Fine-tuning Demo")
    print("=" * 60)

    model = SimpleCNN(num_classes=10)

    # Count initial parameters
    total, nonzero = model.count_parameters()
    print(f"\nInitial model:")
    print(f"  Total parameters: {total:,}")
    print(f"  Non-zero parameters: {nonzero:,}")

    # Simulate training loop with pruning
    print("\nSimulated pruning + fine-tuning schedule:")
    print("-" * 50)

    sparsity_schedule = [0.2, 0.4, 0.6]
    for i, target_sparsity in enumerate(sparsity_schedule):
        print(f"\nIteration {i+1}:")
        print(f"  1. Prune to {target_sparsity*100:.0f}% sparsity")
        print(f"  2. Fine-tune for N epochs")
        print(f"  3. Evaluate accuracy")

    print("\nKey insights:")
    print("  - Pruning removes low-magnitude weights")
    print("  - Fine-tuning recovers accuracy loss")
    print("  - Iterative approach outperforms one-shot")
    print("  - Final model: same architecture, fewer active weights")


# ============================================================================
# Main
# ============================================================================

def main():
    tz.initialize()

    print("=" * 60)
    print("   Model Pruning Example")
    print("=" * 60)

    print("\nComponents demonstrated:")
    print("  Pruning: Magnitude-based, Unstructured, Structured")
    print("  Analysis: Sparsity computation, Layer statistics")
    print("  Training: Iterative pruning with fine-tuning")

    demo_magnitude_pruning()
    demo_layer_wise_pruning()
    demo_iterative_pruning()
    demo_structured_pruning()
    demo_pruning_with_training()

    print("\n" + "=" * 60)
    print("   Pruning example completed successfully!")
    print("=" * 60)


if __name__ == "__main__":
    main()
